partiql-planner/src/main/kotlin/org/partiql/planner/internal/typer/DynamicTyper.kt

```diff
@@ -0,0 +1,323 @@
+                INT to INT,
+                DECIMAL to DECIMAL,
+                DECIMAL_ARBITRARY to DECIMAL_ARBITRARY,
+                FLOAT32 to FLOAT32,
+                FLOAT64 to FLOAT64,
+            )
+            graph[DECIMAL] = edges(
+                INT8 to DECIMAL,
+                INT16 to DECIMAL,
+                INT32 to DECIMAL,
+                INT to DECIMAL,
+                DECIMAL to DECIMAL,
+                DECIMAL_ARBITRARY to DECIMAL_ARBITRARY,
+                FLOAT32 to FLOAT32,
+                FLOAT64 to FLOAT64,
+            )
+            graph[DECIMAL_ARBITRARY] = edges(
+                INT8 to DECIMAL_ARBITRARY,
+                INT16 to DECIMAL_ARBITRARY,
+                INT32 to DECIMAL_ARBITRARY,
+                INT to DECIMAL_ARBITRARY,
+                DECIMAL to DECIMAL_ARBITRARY,
+                DECIMAL_ARBITRARY to DECIMAL_ARBITRARY,
+                FLOAT32 to FLOAT32,
+                FLOAT64 to FLOAT64,
+            )
+            graph[FLOAT32] = edges(
+                INT8 to FLOAT32,
+                INT16 to FLOAT32,
+                INT32 to FLOAT32,
+                INT to FLOAT32,
+                DECIMAL to FLOAT32,
+                DECIMAL_ARBITRARY to FLOAT32,
+                FLOAT32 to FLOAT32,
+                FLOAT64 to FLOAT64,
+            )
+            graph[FLOAT64] = edges(
+                INT8 to FLOAT64,
+                INT16 to FLOAT64,
+                INT32 to FLOAT64,
+                INT to FLOAT64,
+                DECIMAL to FLOAT64,
+                DECIMAL_ARBITRARY to FLOAT64,
+                FLOAT32 to FLOAT64,
+                FLOAT64 to FLOAT64,
+            )
+            graph[CHAR] = edges(
+                CHAR to CHAR,
+                STRING to STRING,
+                SYMBOL to STRING,
+                CLOB to CLOB,
+            )
+            graph[STRING] = edges(
+                CHAR to CHAR,
+                STRING to STRING,
+                SYMBOL to STRING,
+                CLOB to CLOB,
+            )
+            graph[SYMBOL] = edges(
+                CHAR to CHAR,
+                STRING to STRING,
+                SYMBOL to STRING,
+                CLOB to CLOB,
+            )
+            graph[BINARY] = edges(
+                BINARY to BINARY,
+            )
+            graph[BYTE] = edges(
+                BYTE to BYTE,
+                BLOB to BLOB,
+            )
+            graph[BLOB] = edges(
+                BYTE to BLOB,
+                BLOB to BLOB,
+            )
+            graph[DATE] = edges(
+                DATE to DATE,
+            )
+            graph[CLOB] = edges(
+                CHAR to CHAR,
+                STRING to STRING,
+                SYMBOL to STRING,
+                CLOB to CLOB,
+            )
+            graph[TIME] = edges(
+                TIME to TIME,
+            )
+            graph[TIMESTAMP] = edges(
+                TIMESTAMP to TIMESTAMP,
+            )
```

Comment:
SQL:99 also defines some datetime typing rules:

> f) If some data type in DTS is a datetime data type, then every data type in DTS shall be a
datetime data type having the same datetime fields. The result data type is a datetime data
type having the same datetime fields, whose fractional seconds precision is the largest of the
fractional seconds precisions in DTS. If some data type in DTS has a timezone displacement
value, then the result has a timezone displacement value; otherwise, the result does not
have a timezone displacement value.

If we aren't tackling those rules in this PR, could add a TODO